### ISSUE TITLE
input_text doesn't work when using ImString::default()

### ISSUE THREAD

**Issue by bjorn3** [imgui-rs/imgui-rs#366]
Adding any character makes it possible to enter up to 7 characters.

**Comment by bjorn3** [comment]
Using `.reserve(100)` fixes the problem for small inputs.

**Comment by Gekkio** [comment]
It's not obvious but completely expected: ImString is a *manually* growable buffer, and the capacity needs to be managed correctly by the application. This means `ImString::default()` gives you a buffer with enough space for exactly 0 characters.

The upstream Dear ImGui library nowadays provides a resize callback which could be used to implement automatic buffer capacity management, but this is still on my TODO list.

**Comment by bjorn3** [comment]
If I use `.reserve(1)` to grow as necessary before every `ui.input_text()` call it still gets limited to 7 characters.

**Comment by Gekkio** [comment]
It seems that some upstream Dear ImGui library internal implementation details are leaking: the C++ implementation seems to make a defensive copy of the buffer on widget activation, and the capacity of this internal buffer is separate from the capacity of the imgui-rs side buffer (which I verified grows as expected). In practice this means that the maximum string length is fixed when you focus on the text input by clicking. You can try this yourself: when you reach the 7 character limit, try clicking something else, then the text input again, and see how you can enter more characters (7?) until you reach the next limit. You can repeat this multiple times, and moving focus away from the text input always extends the C++ -side buffer by a couple of characters.

In other words: nasty implementation detail that can't be fixed right now by imgui-rs, because we're passing a buffer with enough capacity but Dear ImGui isn't using it. I assume using the resize callback always behind the scenes will fix this.

**Comment by bjorn3** [comment]
That explains it. Thanks!

**Comment by sanbox-irl** [comment]
Going to close this issue. See the update on https://github.com/imgui-rs/imgui-rs/issues/380
